### ISSUE TITLE
problem with user access manager and threewp-broadcast plugin

### ISSUE THREAD

**Issue by kginc** [GM-Alex/user-access-manager#194]
Hi, I am trying to use the user access manager with threewp-broadcast plugin. Unfortunately I get an error message. The developer of the plugin told me, that there maybe is a problem with the switch_to_blog() function (Looking through the source code, it calls switch_to_blog (switchToBlog) several times but not does call restore_current_blog();) Could you please check and - if there is a wrong function - please fix it. Thanks in advance

**Comment by GM-Alex** [comment]
The plugin switches always back to the current blog which we should be the same like `restore_current_blog`, so that's shouldn't be the issue. I will close this until you can provide more detailed information about the issue.

**Comment by plainviewp1206** [comment]
Your plugin makes several calls to switch_to_blog() via switchToBlog(), but does not match them with an equal amount of calls to restore_current_blog().

According to the Wordpress docs, https://codex.wordpress.org/Function_Reference/switch_to_blog, 

> If you do not call restore_current_blog() after every switch_to_blog(), WordPress can get into a state that can potentially build the wrong urls for the site. See restore_current_blog() vs switch_to_blog()

This is preventing Broadcast from working due to the inconsistency.

**Comment by GM-Alex** [comment]
@plainviewp1206 I don't use `restore_current_blog` because it's functional equivalent to

```php
$currentId = $wpdb->blogid;
switch_to_blog($currentId);
```

See for example https://github.com/GM-Alex/user-access-manager/blob/master/src/Controller/Backend/SetupController.php#L111

```php
$currentBlogId = $this->database->getCurrentBlogId();

foreach ($blogIds as $blogId) {
    $this->wordpress->switchToBlog($blogId);
    ...
}

$this->wordpress->switchToBlog($currentBlogId);
```

So it will never match the amount of `restore_current_blog`...

**Comment by plainviewp1206** [comment]
You are incorrect.

https://wordpress.stackexchange.com/questions/89113/restore-current-blog-vs-switch-to-blog

This is from the first answer:

> After every instance of switch_to_blog() you need to call restore_current_blog() otherwise WP will think it is in a "switched" mode and can potentially return incorrect data.

Also, check https://codex.wordpress.org/Function_Reference/ms_is_switched

```
function ms_is_switched()
{
  return ! empty( $GLOBALS['_wp_switched_stack'] );
}
```

Notice that it talks about a **stack**? Your constant switching blogs is just pushing to the stack, never popping, so the stack will never return to an empty state.

**Comment by GM-Alex** [comment]
@plainviewp1206 See my edit...

**Comment by kginc** [comment]
Thank you very much for fixing the problem!



Von: Alexander Schneider [mailto:notifications@github.com]
Gesendet: Freitag, 15. Dezember 2017 21:42
An: GM-Alex/user-access-manager <user-access-manager@noreply.github.com>
Cc: kginc <kg@increon.com>; Author <author@noreply.github.com>
Betreff: Re: [GM-Alex/user-access-manager] problem with user access manager and threewp-broadcast plugin (#194)


Closed #194<https://github.com/GM-Alex/user-access-manager/issues/194> via 4f3af47<https://github.com/GM-Alex/user-access-manager/commit/4f3af47725cf025da570215a2a1c92507d3f7ceb>.
